File: freight-triage-polyglot-ledger-recovery/environment/intake/src/com/freight/hash/Adler32.java
package com.freight.hash;

/** adler32 over raw bytes. */
public final class Adler32 implements HashAlgorithm {

    @Override
    public String name() {
        return "adler32";
    }

    @Override
    public long apply(byte[] data) {
        long low = 1L;
        long high = 0L;
        for (int i = 0; i < data.length; i++) {
            low = (low + (data[i] & 0xFF)) % 65521L;
            high = (high + low) % 65521L;
        }
        return ((high << 16) | low) & 0xFFFFFFFFL;
    }
}

File: freight-triage-polyglot-ledger-recovery/environment/native/src/hash/adler32.cpp
#include "freight/hashes.h"

namespace freight {

// adler32 over raw bytes.
uint64_t hash_adler32(const std::string& data) {
  uint32_t low = 1u;
  uint32_t high = 0u;
  for (size_t i = 0; i < data.size(); ++i) {
    low = (low + static_cast<uint8_t>(data[i])) % 65521u;
    high = (high + low) % 65521u;
  }
  return static_cast<uint64_t>((high << 16) | low);
}

}  // namespace freight
